docs/source/en/tasks/prompting.md

```diff
@@ -0,0 +1,466 @@
+Text generation strategies and parameters are out of scope for this guide, but you can learn more about these topics in 
+the following guides: 
+ 
+* [Generation with LLMs](../llm_tutorial)
+* [Text generation strategies](../generation_strategies)
+
+</Tip>
+
+## Basics of prompting
+
+### Types of models 
+
+The majority of modern LLMs are decoder-only transformers. Some examples include: [LLaMA](../model_doc/llama), 
+[Llama2](../model_doc/llama2), [Falcon](../model_doc/falcon), [GPT2](../model_doc/gpt2). However, you may encounter  
+encoder-decoder transformer LLMs as well, for instance, [Flan-T5](../model_doc/flan-t5) and [BART](../model_doc/bart).
+
+Encoder-decoder-style models are typically used in generative tasks where the output **heavily** relies on the input, for 
+example, in translation and summarization. The decoder-only models are used for all other types of generative tasks.
+
+When using a pipeline to generate text with an LLM, it's important to know what type of LLM you are using, because 
+they use different pipelines. 
+
+Run inference with decoder-only models with the `text-generation` pipeline:
+
+```python
+>>> from transformers import pipeline
+>>> import torch
+
+>>> torch.manual_seed(0) # doctest: +IGNORE_RESULT
+
+>>> generator = pipeline('text-generation', model = 'gpt2')
+>>> prompt = "Hello, I'm a language model"
+
+>>> generator(prompt, max_length = 30)
+[{'generated_text': "Hello, I'm a language model expert, so I'm a big believer in the concept that I know very well and then I try to look into"}]
+```
+
+To run inference with an encoder-decoder, use the `text2text-generation` pipeline:
+
+```python
+>>> text2text_generator = pipeline("text2text-generation", model = 'google/flan-t5-base')
+>>> prompt = "Translate from English to French: I'm very happy to see you"
+
+>>> text2text_generator(prompt)
+[{'generated_text': 'Je suis très heureuse de vous rencontrer.'}]
+```
+
+### Base vs instruct/chat models
+
+Most of the recent LLM checkpoints available on 🤗 Hub come in two versions: base and instruct (or chat). For example, 
+[`tiiuae/falcon-7b`](https://huggingface.co/tiiuae/falcon-7b) and [`tiiuae/falcon-7b-instruct`](https://huggingface.co/tiiuae/falcon-7b-instruct).
+
+Base models are excellent at completing the text when given an initial prompt, however, they are not ideal for NLP tasks 
+where they need to follow instructions, or for conversational use. This is where the instruct (chat) versions come in. 
+These checkpoints are the result of further fine-tuning of the pre-trained base versions on instructions and conversational data. 
+This additional fine-tuning makes them a better choice for many NLP tasks.  
+
+Let's illustrate some simple prompts that you can use with [`tiiuae/falcon-7b-instruct`](https://huggingface.co/tiiuae/falcon-7b-instruct) 
+to solve some common NLP tasks.
+
+### NLP tasks 
+
+First, let's set up the environment: 
+
+```bash
+pip install -q transformers einops accelerate
```

Comment:
```suggestion
pip install -q transformers accelerate
```

We don't need `einops` here no?

Reply:
Unfortunately we do. Without it I get an ImportError:
```ImportError: This modeling file requires the following packages that were not found in your environment: einops. Run `pip install einops```

Reply:
Ah it's because of the trust_remote_code - I see! A bit unfortunate because we've been strongly avoiding `einops` so far, but maybe we could remember to update this code snippet here once Falcon is fully in core transformers (cc @ArthurZucker )

Reply:
Actually I think this can be removed here, as well as the `trust_remote_code=True` parameter. If you remove the `trust_remote_code=True` parameter, we'll leverage the internal implementation by Matt which does not leverage `einops` or remote code.

Verified on `main`:

```
(.env) (base) lysandre@x:~$ pip uninstall einops
Found existing installation: einops 0.6.1
Uninstalling einops-0.6.1:
  Would remove:
    /home/lysandre/.env/lib/python3.11/site-packages/einops-0.6.1.dist-info/*
    /home/lysandre/.env/lib/python3.11/site-packages/einops/*
Proceed (Y/n)?
  Successfully uninstalled einops-0.6.1
(.env) (base) lysandre@x:~$ ipython
Python 3.11.3 (main, May 30 2023, 15:36:33) [GCC 9.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: >>> from transformers import pipeline, AutoTokenizer
   ...: >>> import torch
   ...:
   ...: >>> torch.manual_seed(0) # doctest: +IGNORE_RESULT
   ...: >>> model = "tiiuae/falcon-7b-instruct"
   ...:
   ...: >>> tokenizer = AutoTokenizer.from_pretrained(model)
   ...: >>> pipe = pipeline(
   ...: ...     "text-generation",
   ...: ...     model=model,
   ...: ...     tokenizer=tokenizer,
   ...: ...     torch_dtype=torch.bfloat16,
   ...: ...     device_map="auto",
   ...: ... )
Downloading (…)okenizer_config.json: 100%|███████████████████████████████████████████████████| 220/220 [00:00<00:00, 1.64MB/s]
Downloading (…)/main/tokenizer.json: 100%|███████████████████████████████████████████████| 2.73M/2.73M [00:00<00:00, 35.3MB/s]
Downloading (…)cial_tokens_map.json: 100%|███████████████████████████████████████████████████| 281/281 [00:00<00:00, 2.60MB/s]
Loading the tokenizer from the `special_tokens_map.json` and the `added_tokens.json` will be removed in `transformers 5`,  it is kept for forward compatibility, but it is recommended to update your `tokenizer_config.json` by uploading it again. You will see the new `added_tokens_decoder` attribute that will store the relevant information.
The Falcon model was initialized without `trust_remote_code=True`, and will therefore leverage the transformers library implementation. tiiuae/falcon-7b-instruct's revision is set to a version that doesn't leverage remote code (f8dac3fff96d5debd43edf56fb4e1abcfffbef28).

In order to override this, please set a revision manually or set `trust_remote_code=True`.
Downloading (…)fffbef28/config.json: 100%|███████████████████████████████████████████████| 1.05k/1.05k [00:00<00:00, 8.50MB/s]
The Falcon model was initialized without `trust_remote_code=True`, and will therefore leverage the transformers library implementation. tiiuae/falcon-7b-instruct's revision is set to a version that doesn't leverage remote code (f8dac3fff96d5debd43edf56fb4e1abcfffbef28).

In order to override this, please set a revision manually or set `trust_remote_code=True`.
Downloading (…)model.bin.index.json: 100%|████████████████████████████████████████████████| 16.9k/16.9k [00:00<00:00, 107MB/s]
Downloading (…)l-00001-of-00002.bin: 100%|████████████████████████████████████████████████| 9.95G/9.95G [00:18<00:00, 544MB/s]
Downloading (…)l-00002-of-00002.bin: 100%|████████████████████████████████████████████████| 4.48G/4.48G [00:15<00:00, 283MB/s]
Downloading shards: 100%|███████████████████████████████████████████████████████████████████████| 2/2 [00:34<00:00, 17.09s/it]
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████| 2/2 [00:06<00:00,  3.20s/it]
Downloading (…)neration_config.json: 100%|████████████████████████████████████████████████████| 117/117 [00:00<00:00, 766kB/s]

In [2]:
```

Reply:
Awesome! Removed the `trust_remote_code=True` parameter and einops.